llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -2275,18 +2275,36 @@ SDValue DAGCombiner::visitPTRADD(SDNode *N) {
     SDValue Z = N1;
 
     SDValue Add = DAG.getNode(ISD::ADD, DL, IntVT, {Y, Z});
+    // Calling visit() can replace the Add node with ISD::DELETED_NODE if there
+    // aren't any users. Therefore, we created the reassociated result now
+    // regardless of whether we end up returning it since otherwise we might end
+    // up using an invalid node in the checks below.
+    SDValue Reassociated = DAG.getPointerAdd(DL, X, Add);
     SDValue VisitedAdd = visit(Add.getNode());
-    if (VisitedAdd)
-      Add = VisitedAdd;
-
+    if (VisitedAdd) {
+      // If visit() returns the same node, it means the SDNode was RAUW'd, and
+      // therefore we have to load the new value to perform the checks whether
+      // the reassociation fold is profitable. Otherwise just update Add to the
+      // simplified node.
+      if (VisitedAdd.getNode() == Add.getNode())
+        Add = Reassociated.getOperand(1);
+      else
+        Add = VisitedAdd;
+    }
+    LLVM_DEBUG(dbgs() << "visitPTRADD() add operand:"; Add.dump(&DAG));
+    LLVM_DEBUG(dbgs() << "visitPTRADD() reassociated:";
+               Reassociated.dump(&DAG));
+    assert(Add->getOpcode() != ISD::DELETED_NODE && "Deleted Node used");
     if (isNullConstant(X) ||
         DAG.isConstantIntBuildVectorOrConstantInt(Add) ||
         (VisitedAdd && !DAG.isConstantIntBuildVectorOrConstantInt(Z)) ||
         (DAG.isConstantIntBuildVectorOrConstantInt(Y) && Z.hasOneUse()) ||
         (DAG.isConstantIntBuildVectorOrConstantInt(Y) && N0.hasOneUse()) ||
         (N0.hasOneUse() && Z.hasOneUse() &&
          !DAG.isConstantIntBuildVectorOrConstantInt(Z))) {
-      return DAG.getPointerAdd(DL, X, Add);
+      return Reassociated;
+    } else {
+      deleteAndRecombine(Reassociated.getNode());
```

Comment:
Seems like calling deleteAndRecombine results in a crash when compiling some other code, will try to reduce and add as a regression test.

Reply:
Fixed now and added a regression test.